### PR TITLE
feat(business_profile_v2): business profile v2 create and retrieve endpoint

### PR DESCRIPTION
**Pull request by Narayanbhat166**
## Type of Change

- [x] New feature

## Description

This PR adds support for the v2 apis of business profile. The endpoints that are supported are 
- business profile create
- business profile retrieve

### Additional Changes

- [x] This PR modifies the API contract
Add business profile create v2 endpoint

## How did you test it?

These tests are conducted locally only. These endpoints are not available on the sandbox environment

- business profile create using the below curl.
```bash
curl --location 'http://localhost:8080/v2/profiles' \
--header 'x-merchant-id: cloth_seller_wIIUnPMwVPUtFq8tbLbd' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
    "profile_name":"My First Profile"
}'
```

- retrieve the business profile.
```bash
curl --location 'http://localhost:8080/v2/profiles/pro_iOBJ44KvGafEJBd6zd7V' \
--header 'x-merchant-id: cloth_seller_wIIUnPMwVPUtFq8tbLbd' \
--header 'api-key: test_admin'
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

File: crates/router/src/core/admin.rs [hunk 1]
```diff
@@ -3311,23 +3309,261 @@ pub async fn create_and_insert_business_profile(
...
+        let current_time = date_time::now();
+
+        let webhook_details = self.webhook_details.map(ForeignInto::foreign_into);
+
+        let payment_response_hash_key = self
+            .payment_response_hash_key
+            .or(merchant_account.payment_response_hash_key.clone())
+            .unwrap_or(common_utils::crypto::generate_cryptographically_secure_random_string(64));
+
+        let payment_link_config_value = self.payment_link_config.map(ForeignInto::foreign_into);
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:3381]: Nit: This is no longer a `serde_json::Value`, I believe this is a remnant from #5531.

```suggestion
        let payment_link_config = self.payment_link_config.map(ForeignInto::foreign_into);
```

File: crates/router/src/core/admin.rs [hunk 2]
```diff
@@ -3311,23 +3309,261 @@ pub async fn create_and_insert_business_profile(
...
+
+        let current_time = date_time::now();
+
+        let webhook_details = self.webhook_details.map(ForeignInto::foreign_into);
+
+        let payment_response_hash_key = self
+            .payment_response_hash_key
+            .unwrap_or(common_utils::crypto::generate_cryptographically_secure_random_string(64));
+
+        let payment_link_config_value = self.payment_link_config.map(ForeignInto::foreign_into);
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:3488]: Same here.

File: crates/router/src/routes/admin.rs [hunk 3]
```diff
@@ -271,6 +277,20 @@ impl<'a> HeaderMapStruct<'a> {
...
+    pub fn get_merchant_id_from_header(
+        &self,
+    ) -> crate::errors::RouterResult<common_utils::id_type::MerchantId> {
+        self.get_mandatory_header_value_by_key(headers::X_MERCHANT_ID.into())
+            .map(|val| val.to_owned())
+            .and_then(|merchant_id| {
+                common_utils::id_type::MerchantId::wrap(merchant_id)
+                    .change_context(ApiErrorResponse::InternalServerError)
+                    .attach_printable(
+                        "Error while converting MerchantId from merchant_id string header",
```
Comment (SanchithHegde) [inline crates/router/src/routes/admin.rs:290]: Nit:

```suggestion
                        "Error while converting MerchantId from `x-merchant-id` string header",
```

File: justfile [hunk 4]
```diff
@@ -63,6 +63,22 @@ check_v2 *FLAGS:
...
 
+run_v2 *FLAGS:
+    #! /usr/bin/env bash
+    set -euo pipefail
+
+    FEATURES="$(cargo metadata --all-features --format-version 1 | \
+        jq -r '
+            [ ( .workspace_members | sort ) as $package_ids # Store workspace crate package IDs in `package_ids` array
+            | .packages[] | select( IN(.id; $package_ids[]) ) | .features | keys[] ] | unique # Select all unique features from all workspace crates
+            | del( .[] | select( any( . ; . == ("v1") ) ) ) # Exclude some features from features list
```
Comment (SanchithHegde) [inline justfile:74]: We can't do this, since it would still include features like AWS KMS, key manager service, etc.

File: justfile [hunk 5]
```diff
@@ -63,6 +63,22 @@ check_v2 *FLAGS:
...
+            | join(",") # Construct a comma-separated string of features for passing to `cargo`
+    ')"
+
+    set -x
+    cargo run --features "${FEATURES}"  {{ FLAGS }}
+    set +x
```
Comment (SanchithHegde) [inline justfile:80]: I'm not sure we need (or can use) FLAGS here, if this recipe only allows running `router` with v2 features.

```suggestion
    FEATURES="$(cargo metadata --all-features --format-version 1 | \
        jq -r '
            [ .packages[] | select(.name == "router") | .features | keys[] # Obtain features of `router` package
            | select( any( . ; test("(([a-z_]+)_)?v2") ) ) ] # Select v2 features
            | join(",") # Construct a comma-separated string of features for passing to `cargo`
    ')"

    set -x
    cargo run --package router --features "${FEATURES}"
    set +x
```

File: v2_migrations/2024-07-31-100300_business_profile_add_v2_columns_drop_v1_columns/up.sql [hunk 6]
```diff
@@ -1,21 +1,17 @@
...
 
+ALTER TABLE business_profile
+ADD COLUMN routing_algorithm_id VARCHAR(64) DEFAULT NULL,
+  ADD COLUMN order_fulfillment_time BIGINT DEFAULT NULL,
+  ADD COLUMN order_fulfillment_time_origin "OrderFulfillmentTimeOrigin" DEFAULT NULL,
+  ADD COLUMN frm_routing_algorithm_id VARCHAR(64) DEFAULT NULL,
+  ADD COLUMN payout_routing_algorithm_id VARCHAR(64) DEFAULT NULL,
+  ADD COLUMN default_fallback_routing JSONB DEFAULT NULL;
+
+-- ADD COLUMN default_fallback_routing JSONB DEFAULT NULL;
```
Comment (SanchithHegde) [inline v2_migrations/2024-07-31-100300_business_profile_add_v2_columns_drop_v1_columns/up.sql:11]: Remove this line.

File: crates/api_models/src/admin.rs [hunk 7]
```diff
@@ -1936,6 +1943,93 @@ pub struct BusinessProfileCreate {
...
+#[derive(Clone, Debug, Deserialize, ToSchema, Default, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct BusinessProfileCreate {
+    /// The name of business profile
+    #[schema(max_length = 64)]
+    pub profile_name: Option<String>,
+
+    /// The URL to redirect after the completion of the operation
+    #[schema(value_type = Option<String>, max_length = 255, example = "https://www.example.com/success")]
+    pub return_url: Option<url::Url>,
```
[the remark this answers is not in the document]
Comment (Narayanbhat166) [reply crates/api_models/src/admin.rs:1962]: I think we should have restriction in the database too, or else someone could abuse this field. Normally the return url should not have more than 255 characters

File: crates/common_utils/src/errors.rs [hunk 8]
```diff
@@ -50,6 +50,9 @@ pub enum ParsingError {
     /// Failed to parse String value to Decimal value conversion because `error`
     #[error("Failed to parse String value to Decimal value conversion because {error}")]
     StringToDecimalConversionFailure { error: String },
+    /// Failed to convert i64 value to u32 value
+    #[error("Failed to parse i64 value to u32 value")]
+    I64ToU32ConversionFailure,
```
Comment (SanchithHegde) [inline crates/common_utils/src/errors.rs:55]: Where is this enum variant being used?

File: crates/router/src/core/admin.rs [hunk 9]
```diff
@@ -3311,23 +3309,261 @@ pub async fn create_and_insert_business_profile(
...
+            collect_billing_details_from_wallet_connector: self
+                .collect_billing_details_from_wallet_connector
+                .or(Some(false)),
+            outgoing_webhook_custom_http_headers: outgoing_webhook_custom_http_headers
+                .map(Into::into),
+        })
+    }
+
+    #[cfg(all(feature = "v2", feature = "business_profile_v2"))]
+    async fn create_domain_model_from_request(
```
Comment (SanchithHegde) [inline crates/router/src/core/admin.rs:3465]: Don't we need to validate order fulfillment time here, after validating session expiry?

```rust
if let Some(order_fulfillment_expiry) = self.order_fulfillment_time {
    helpers::validate_intent_fulfillment_expiry(order_fulfillment_expiry)?;
}
```

Comment (Narayanbhat166) [reply crates/router/src/core/admin.rs:3465]: This validation has been moved to new type

**Review by jarnura — APPROVED** [review]

**Review by Aprabhat19 — APPROVED** [review]

File: crates/router/src/types/api/admin.rs [hunk 10]
```diff
@@ -187,24 +183,22 @@ impl ForeignTryFrom<domain::BusinessProfile> for BusinessProfileResponse {
             })
             .transpose()?;
 
+        let order_fulfillment_time = item
+            .order_fulfillment_time
+            .map(|time| api_models::admin::OrderFulfillmentTime::new(time))
+            .transpose()
+            .change_context(errors::ParsingError::IntegerOverflow)?;
```
Comment (SanchithHegde) [inline crates/router/src/types/api/admin.rs:190]: It's not exactly integer overflow, right? It's that the `order_fulfillment_time` is not within the range 0 to 1800?

**Pull request merged**